src/java/org/apache/cassandra/io/compress/ZSTDCompressor.java

```diff
@@ -0,0 +1,138 @@
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.io.compress;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.github.luben.zstd.Zstd;
+
+public class ZSTDCompressor implements ICompressor
+{
+    public static final int FAST_COMPRESSION = 1; // fastest compression time
+    public static final int DEFAULT_LEVEL = 3;
+    public static final int BEST_COMPRESSION = 22;// very good compression ratio
+    private static final ZSTDCompressor instance = new ZSTDCompressor();
+    private static final String COMPRESSION_LEVEL_OPTION_NAME = "compression_level";
+    @VisibleForTesting
+    protected static final int compressionLevel = DEFAULT_LEVEL;
+
+    public static ZSTDCompressor create(Map<String, String> compressionOptions)
+    {
+        validateCompressionLevel(parseCompressionLevelOption(compressionOptions));
+        return instance;
+    }
+
+    private static void validateCompressionLevel(int compressionLevel)
+    {
+        if (compressionLevel < FAST_COMPRESSION || compressionLevel > BEST_COMPRESSION)
+        {
+            throw new IllegalArgumentException(
+                "ZSTD compression_level " + Integer.toString(compressionLevel) + " invalid ",
+                null
+            );
+        }
+    }
+
+    private static int parseCompressionLevelOption(Map<String,String> compressionOptions)
+    {
+        return Integer.parseInt(compressionOptions.getOrDefault(COMPRESSION_LEVEL_OPTION_NAME,
+                                                                Integer.toString(DEFAULT_LEVEL)));
+    }
+
+    @Override
+    public int initialCompressedBufferLength(int chunkLength)
+    {
+        return (int)Zstd.compressBound(chunkLength);
+    }
+
+    @Override
+    public int uncompress(byte[] input, int inputOffset, int inputLength, byte[] output, int outputOffset)
+    throws IOException
+    {
+        long decompressSize = Zstd.decompressByteArray(output,
+                                                       outputOffset,
+                                                       output.length - outputOffset,
+                                                       input,
+                                                       inputOffset,
+                                                       inputLength);
+        if (Zstd.isError(decompressSize))
+        {
+            throw new IOException("ZSTD uncompress failed with error reason " + Zstd.getErrorName(decompressSize));
+        }
+
+        return (int) decompressSize;
+    }
+
+    @Override
+    public void uncompress(ByteBuffer input, ByteBuffer output) throws IOException
+    {
+        Zstd.decompress(output, input);
+    }
+
+    @Override
+    public void compress(ByteBuffer input, ByteBuffer output) throws IOException
```

Comment:
I think you may be able to replace this with the [``compress(ByteBuffer dstBuf, ByteBuffer srcBuf, int level)``](https://github.com/luben/zstd-jni/blob/master/src/main/java/com/github/luben/zstd/Zstd.java#L535) method, you just have to make sure to pass the arguments in the opposite order (output comes first) and re-throw the `RuntimeException` as an `IOException`.

Reply:
ICompressor Interface supports only compress(ByteBuffer input, ByteBuffer output)

Reply:
Indeed, but I think you can simplify this implementation by using the built in `Zstd.compress` method I linked about which afaict does what the code here does.